### ISSUE TITLE
Inconsistent arg parsing behaviour in cli

### ISSUE THREAD

**Issue by MatthewCroughan** [foundriesio/fioctl#44]
`fioctl devices help` works.
`fioctl devices config help` works
`fioctl devices config set help` does not work

Instead, when you get to this level, you need to provide `-h` instead of `help`. Like this: `fioctl devices config set -h`

**Comment by doanac** [comment]
wow - i'm surprised placing a raw `help` works at all. My understanding of the arg parsing library was that you always do `--help`. I see that's not right now.

**Comment by doanac** [comment]
i see the difference now. "help" can work if something has subcommands. But if the option is an actual command  "help" wouldn't work because then it could be the argument to the command. ie - if you have a device named "help", `fioctl devices config set help` actually means something.

I think for the purpose of documentation we just say "--help" to get people in a consistent mode. However, `fioctl devices config` doesn't claim it has a "help" subcommand. So I fear this might be an awkward thing to hack around in the arg parsing library.

**Comment by MatthewCroughan** [comment]
`fioctl devices -f <factory> list` works 
`fioctl devices list -f <factory>` works

The latter is the correct ordering, though both work. May not be an inconsistency, just found it strange.

**Comment by doanac** [comment]
I'm not sure I agree that any of this is inconsistent or strange. Examples include:
~~~
 # coreutils
 rm -rf /tmp/foo 
 rm /tmp/foo -rf 

 # kubectl
 kubectl --namespace default get pods
 kubectl get pods --namespace default
~~~

Good CLI parsers (like [Viper](https://github.com/spf13/viper) which we use) also do:
~~~
 --factory foo
 --factory=foo
~~~

And `help` is only available to sub-commands (ie `fioctl devices`) and not *commands* `fioctl devices list`

**Comment by MatthewCroughan** [comment]
That's interesting, I did not know there were multiple ways to layout the position of flags in coreutils. 

I have only used some modern apps that allow the usage of `help` without --help. Though they do not allow `help` without `--` in their subcommands, which seems pointless to have allowed it in the first place. 

```
matthew@thinkpad ~ $ wormhole help
Usage: wormhole [OPTIONS] COMMAND [ARGS]...

  Create a Magic Wormhole and communicate through it.

  Wormholes are created by speaking the same magic CODE in two different
  places at the same time.  Wormholes are secure against anyone who doesn't
  use the same code.

Options:
  --appid APPID                   appid to use
  --relay-url URL                 rendezvous relay to use
  --transit-helper tcp:HOST:PORT  transit relay to use
  --dump-timing FILE.json         (debug) write timing data to file
  --version                       Show the version and exit.
  --help                          Show this message and exit.

Commands:
  help
  receive  Receive a text message, file, or directory (from 'wormhole send')
  send     Send a text message, file, or directory
  ssh      Facilitate sending/receiving SSH public keys

matthew@thinkpad ~ $ wormhole ssh help
Usage: wormhole ssh [OPTIONS] COMMAND [ARGS]...
Try 'wormhole ssh --help' for help.

Error: No such command 'help'.

matthew@thinkpad ~ $ wormhole ssh --help
Usage: wormhole ssh [OPTIONS] COMMAND [ARGS]...

  Facilitate sending/receiving SSH public keys

Options:
  --help  Show this message and exit.

Commands:
  accept  Send your SSH public-key In response to a 'wormhole ssh invite'...
  invite  Add a public-key to a ~/.ssh/authorized_keys file

```

The only thing I am worried about is that people get used to typing `help`, finding it works, then finding it does not work if given to a command as I did.

This is a small issue anyway, I do not think it will matter if we simply tell people to use -h in the docs, as I am already doing.

**Comment by MatthewCroughan** [comment]
The flag behaviour I commented on as 'strange' is just my lack of knowledge and experience. I still believe the allowance of `help` for subcommands instead of allowing only `--help` or `-h` like everything else introduces some inconsistency that leads to confusion though.

**Comment by doanac** [comment]
Possibly, but there's lots of good tools doing this same thing:

  *  Hugo
  *  EMC RexRay
  *  Imgur’s Incus
  *  Nanobox/Nanopack
  *  Docker Notary
  *  BloomApi
  *  doctl
  *  Clairctl
  *  Mercure
  * kubectl

I think being like them instead of trying to hack Viper to *not* do this makes the most sense.

**Comment by MatthewCroughan** [comment]
Ah I see. I will play with some of those tools and see if they end up with the same behavior. I plan on using Hugo myself. Though I agree, if this does require hacking the library, it is not at all worth it, it is just best that we tell people that `-h` is the way of doing things, and if they accidentally discover as `help` I did then so what.

The main reason I used `help` instead of `-h` is because it was listed first as the output of `fioctl --help` as an "available command". So I just stuck with that throughout my usage.
